<launch>

    <!-- this launch file is used to test test backprojection of panorama camera.. -->


    <!-- general parameters -->
    <arg name="camera_frame" default="/hardware/video/valeo/ValeoLeftImageFrame" />
    <arg name="world_frame" default="/robot/OdometryFrame" />

    <!-- GP -->
    <arg name="rate" default="10.0"/>
    <arg name="base_footprint" default="/robot/BaseFrame" />
    <arg name="ground_plane" default="/ground_plane" />


   <!-- Ground Plane (world) -->
    <node pkg="rwth_ground_plane" type="ground_plane_tf_based_fixed" name="ground_plane" output="screen">
        <param name="base_footprint" value="$(arg base_footprint)" />
        <param name="camera_frame" value="$(arg world_frame)" />
        <param name="ground_plane" value="$(arg ground_plane)" />
        <param name="rate" value="$(arg rate)"/>
    </node>




    <!-- general yoloconvertor_panorama arg-->
    <arg name="detection_id_offset" default="0"/>
    <arg name="detection_id_increment" default="1"/>
    <arg name="world_scale" default="1"/>
    <arg name="queue_size" default="10" />

    <!-- left yoloconvertor and yolo arg-->
    <arg name="pano_image_left" default="/hardware/video/valeo/rectificationNIKRLeft/PanoramaImage" />
    <arg name="detected_persons_left" default="/yoloconvertor_pano/detected_persons_left"/>
    <arg name="bounding_boxes_left" default="darknet_ros/bounding_boxes_left"/>


    <!-- left yoloconvertor and yolo-->
    <!-- Start darknet and ros wrapper -->
    <!-- Load parameters for left -->
   <arg name="engine_path" default="$(find trtyolo)/tensorRT_yolo/yolov3_fp16.engine"/>
   <arg name="detect_threshold" default="0.7"/>
   <arg name="nms_threshold" default="0.45"/>
   <arg name="tensorRT_yolo_out_image" default="/tensorRT_yolo_image"/>
   <node pkg="trtyolo" type="tensorrt_yolo" name= "tensorrtyolo" output="screen">
       <param name="engine_path" value="$(arg engine_path)" type="string"/>
       <param name="detect_threshold" value="$(arg detect_threshold)"/>
       <param name="nms_threshold" value="$(arg nms_threshold)"/>
       <param name="bounding_boxes" value="$(arg bounding_boxes_left)" type="string"/>
       <param name="image" value="$(arg pano_image_left)" type="string"/>
       <param name="tensorRT_yolo_out_image" value="$(arg tensorRT_yolo_out_image)" type="string"/>
   </node>	


    <arg name="occupancy_threshold" default="75.0"/>
    <arg name="occupancy_check_box_half_length" default="1"/>
    <arg name="map" default="/map"/>
    <node pkg="yoloconvertor" type="yoloconvertor_panorama" name="yoloconvertor_panorama_left" output="screen"> 
        <param name="world_scale" value="$(arg world_scale)"/>
        <param name="ground_plane" value="$(arg ground_plane)" type="string"/>
        <param name="queue_size" value="$(arg queue_size)" type="int"/>
        <param name="pano_image" value="$(arg pano_image_left)" type="string"/>
        <param name="detected_persons" value="$(arg detected_persons_left)" type="string"/>
        <param name="detection_id_offset" value="$(arg detection_id_offset)"/>
        <param name="detection_id_increment" value="$(arg detection_id_increment)"/>
        <param name="bounding_boxes" value="$(arg bounding_boxes_left)"/>
	<param name="occupancy_threshold" value="$(arg occupancy_threshold)" type="double"/> <!--from 0 to 100-->
        <param name="occupancy_check_box_half_length" value="$(arg occupancy_check_box_half_length)" type="int"/>
        <param name="map" value="$(arg map)" type="string"/>
    </node>



   <!-- arg for yolofusion-->
   <arg name="fused_detections" default="yolofusion/detected_persons" />
   <!-- run yolofusion-->
   <node pkg="yoloconvertor" type="yoloconvertor_fusion" name="fuse_yolo" output="screen">
       <param name="world_scale" value="$(arg world_scale)"/>
       <param name="queue_size" value="$(arg queue_size)" type="int"/>
       <param name="detection_id_offset" value="0"/>
       <param name="detection_id_increment" value="1"/>
       <param name="pose_variance" value="0.05"/>
       <param name="threshold" value="0.05"/>
       <param name="total_detected_persons" value="$(arg fused_detections)" type="string"/>
       <param name="detected_persons_left" value="$(arg detected_persons_left)" type="string"/>
       <param name="detected_persons_right" value="$(arg detected_persons_left)" type="string"/>
       <param name="detected_persons_rear" value="$(arg detected_persons_left)" type="string"/>
       <param name="left_camera_frame" value="/hardware/video/valeo/ValeoLeftImageFrame" type="string"/>
       <param name="right_camera_frame" value="/hardware/video/valeo/ValeoLeftImageFrame" type="string"/>
       <param name="rear_camera_frame" value="/hardware/video/valeo/ValeoLeftImageFrame" type="string"/>
       <param name="world_frame" value="$(arg world_frame)" type="string"/>
   </node>



    <!-- VO -->
    <arg name="motion_parameters" default="/motion_matrix" />
    <arg name="sensor_frame_id" default="/hardware/depth/kinect2/ColorImageFrame" />
    <arg name="odom_frame_id" default="/robot/OdometryFrame" />
    <!-- Visual Odometry -->
    <include file="$(find odometry_to_motion_matrix)/launch/tf2visual.launch">
        <arg name="motion_parameters" value="$(arg motion_parameters)" />
        <arg name="sensor_frame_id" value="$(arg sensor_frame_id)" />
        <arg name="odom_frame_id" value="$(arg odom_frame_id)" />
    </include>

    <!-- tracker -->
    <arg name="tracked_persons" default="/rwth_tracker/tracked_persons" />
    <arg name="camera_namespace" default="/hardware/depth/kinect2" />
    <!-- Pedestrian Tracking -->
    <include file="$(find rwth_pedestrian_tracking)/launch/pedestrian_tracking.launch">
        <arg name="camera_namespace" value="$(arg camera_namespace)" />
        <arg name="ground_plane" value="$(arg ground_plane)" />
        <arg name="detections" value="$(arg fused_detections)" />
        <arg name="visual_odometry" value="$(arg motion_parameters)" />
        <arg name="pedestrian_array" value="/rwth_tracker/pedestrian_array" />
        <arg name="pedestrian_image" value="/rwth_tracker/image" />
        <arg name="tracked_persons" value="$(arg tracked_persons)" />
        <arg name="queue_size" value="30" />
    </include>





	    <!-- trajectories -->
    <arg name="person_trajectories" default="/rwth_tracker/person_trajectories" />
    <arg name="selected_person_trajectory" default="/rwth_tracker/selected_person_trajectory" />
    <!-- Person Trajectories -->
    <include file="$(find rwth_pedestrian_trajectories)/launch/pedestrian_trajectories.launch">
        <arg name="tracked_persons" value="$(arg tracked_persons)" />
        <arg name="person_trajectories" value="$(arg person_trajectories)" />
        <arg name="selected_person_trajectory" value="$(arg selected_person_trajectory)" />
        <arg name="queue_size" value="10" />
        <arg name="camera_frame" value="$(arg camera_frame)" />
    </include>


    <!-- arg of back projection -->
    <arg name="pano_image" default="/hardware/video/valeo/rectificationNIKRLeft/PanoramaImage" />
    <arg name="node_name" default="back_project_to_left_camera" />
    <arg name="tracked_person_2d" default="tracked_person_2d" />
    <arg name="back_project_visul_image" default="back_project_visul_image" />
 

   <!-- run back projection-->
   <node pkg="yoloconvertor" type="backprojection_panorama" name= "$(arg node_name)" output="screen">
       <param name="pano_image" value="$(arg pano_image)" type="string"/>
       <param name="queue_size" value="$(arg queue_size)" type="int"/>
       <param name="tracked_persons" value="$(arg tracked_persons)" type="string"/>
       <param name="camera_frame" value="$(arg camera_frame)" type="string"/>
       <param name="world_frame" value="$(arg world_frame)" type="string"/>
       <param name="tracked_person_2d" value="$(arg tracked_person_2d)" type="string"/>
       <param name="back_project_visul_image" value="$(arg back_project_visul_image)" type="string"/>
        <param name="person_trajectories" value="$(arg person_trajectories)" />
   </node>


   <!-- run map-->
   <arg name="map_path"          default="$(find yoloconvertor)/map/first_floor.yaml"/>
   <arg name="map_frame_id"  default="/maps/Zuse/Floor1MapFrame" />
   <node pkg="map_server" type="map_server" name= "global_map" args="$(arg map_path)" output="screen" >
       <param name="frame_id" value="$(arg map_frame_id)" type="string"/>
   </node>	

    <!-- for test!!! -->
    <arg name="use_sim_time" default="true" />
    <!-- adding rostopic tool relay -->
    <node name="topic_relay3" pkg="topic_tools" type="relay" args="/hardware/depth/kinect2/ColorIntrinsic /hardware/depth/kinect2/hd/camera_info" respawn="true" />
    <!--
    <node name="topic_relay5" pkg="topic_tools" type="relay" args="/hardware/depth/kinect2/sync/SyncRegisteredDepthImage /hardware/depth/kinect2/hd/image_depth_rect" respawn="true" />
    -->
    <node name="topic_relay2" pkg="topic_tools" type="relay" args="/hardware/depth/kinect2/sync/SyncColorImage /hardware/depth/kinect2/hd/image_color_rect" respawn="true" />	
</launch> 	
